pkg/dockerregistry/server/tagservice.go

```diff
@@ -0,0 +1,187 @@
+		context.GetLogger(ctx).Errorf("Error retrieving ImageStream %s/%s: %v", t.repo.namespace, t.repo.name, err)
+		return distribution.Descriptor{}, distribution.ErrRepositoryUnknown{Name: t.repo.Named().Name()}
+	}
+
+	te := imageapi.LatestTaggedImage(imageStream, tag)
+	if te == nil {
+		return distribution.Descriptor{}, distribution.ErrTagUnknown{Tag: tag}
+	}
+	dgst, err := digest.ParseDigest(te.Image)
+	if err != nil {
+		return distribution.Descriptor{}, err
+	}
+
+	if !t.repo.pullthrough {
+		image, err := t.repo.getImage(dgst)
+		if err != nil {
+			return distribution.Descriptor{}, err
+		}
+
+		if !isImageManaged(image) {
+			return distribution.Descriptor{}, distribution.ErrTagUnknown{Tag: tag}
+		}
+	}
+
+	return distribution.Descriptor{Digest: dgst}, nil
+}
+
+func (t tagService) All(ctx context.Context) ([]string, error) {
+	tags := []string{}
+
+	imageStream, err := t.repo.getImageStream()
+	if err != nil {
+		context.GetLogger(ctx).Errorf("Error retrieving ImageStream %s/%s: %v", t.repo.namespace, t.repo.name, err)
+		return tags, distribution.ErrRepositoryUnknown{Name: t.repo.Named().Name()}
+	}
+
+	for tag, history := range imageStream.Status.Tags {
+		if len(history.Items) == 0 {
+			continue
+		}
+		if !t.repo.pullthrough {
+			dgst, err := digest.ParseDigest(history.Items[0].Image)
+			if err != nil {
+				return nil, err
+			}
+
+			image, err := t.repo.getImage(dgst)
+			if err != nil {
+				return nil, err
+			}
+
+			if !isImageManaged(image) {
+				continue
+			}
+		}
+		tags = append(tags, tag)
+	}
+	return tags, nil
+}
+
+func (t tagService) Lookup(ctx context.Context, desc distribution.Descriptor) ([]string, error) {
+	tags := []string{}
+
+	imageStream, err := t.repo.getImageStream()
+	if err != nil {
+		context.GetLogger(ctx).Errorf("Error retrieving ImageStream %s/%s: %v", t.repo.namespace, t.repo.name, err)
+		return tags, distribution.ErrRepositoryUnknown{Name: t.repo.Named().Name()}
+	}
+
+	for tag, history := range imageStream.Status.Tags {
+		if len(history.Items) == 0 {
+			continue
+		}
+
+		dgst, err := digest.ParseDigest(history.Items[0].Image)
+		if err != nil {
+			return tags, err
+		}
+
+		if !t.repo.pullthrough {
+			image, err := t.repo.getImage(dgst)
+			if err != nil {
+				return nil, err
+			}
+
+			if !isImageManaged(image) {
+				continue
+			}
+		}
+
+		if dgst == desc.Digest {
+			tags = append(tags, tag)
+		}
+	}
+
+	return tags, nil
+}
+
+func (t tagService) Tag(ctx context.Context, tag string, dgst distribution.Descriptor) error {
+	imageStream, err := t.repo.getImageStream()
+	if err != nil {
+		context.GetLogger(ctx).Errorf("Error retrieving ImageStream %s/%s: %v", t.repo.namespace, t.repo.name, err)
+		return distribution.ErrRepositoryUnknown{Name: t.repo.Named().Name()}
+	}
+
+	image, err := t.repo.registryOSClient.Images().Get(dgst.Digest.String())
+	if err != nil {
+		context.GetLogger(ctx).Errorf("(tagService).Tag: Unable to get image: %s", dgst.Digest.String())
+		return err
```

Comment:
Here I can return _this_ error here. `docker/distribution` passes the error from the driver. It can be any error.

https://github.com/docker/distribution/blob/master/registry/storage/tagstore.go#L88